File: jetbrains-plugin/src/main/kotlin/com/vscode/jetbrainssync/MulticastManager.kt
package com.vscode.jetbrainssync

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import java.net.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.thread

// ==================== Internal Data Classes ====================


/**
 * Multicast Manager
 * Responsible for sending and receiving UDP multicast messages, implementing decentralized editor synchronization
 */
class MulticastManager(
    private val project: Project,
    private val messageProcessor: MessageProcessor,
) {
    private val log: Logger = Logger.getInstance(MulticastManager::class.java)

    // ==================== Configuration Constants ====================
    private val multicastAddress = "224.0.0.1" // Local link multicast address, local machine communication only
    private var multicastPort: Int // Multicast port (read from configuration)
    private val maxMessageSize = 8192 // Maximum message size (8KB)

    // ==================== Network Components ====================
    private var multicastSocket: MulticastSocket? = null
    private var networkInterface: NetworkInterface? = null
    private var group: InetSocketAddress? = null

    // ==================== State Management ====================
    private val connectionState = AtomicReference(ConnectionState.DISCONNECTED)
    private val autoReconnect = AtomicBoolean(false)
    private var connectionCallback: ConnectionCallback? = null

    // ==================== Thread Management ====================
    private val executorService: ExecutorService = Executors.newCachedThreadPool { r ->
        val thread = Thread(r, "Multicast-Manager-Worker")
        thread.isDaemon = true
        thread
    }
    private val connectionLock = ReentrantLock()
    private val isShutdown = AtomicBoolean(false)
    private var receiverThread: Thread? = null

    init {
        // Read multicast port from configuration
        val settings = VSCodeJetBrainsSyncSettings.getInstance(project)
        multicastPort = if (settings.state.useCustomPort) {
            settings.state.customPort
        } else {
            3000 // Default port for multicast when auto-detection is used
        }
        log.info("Initializing multicast manager - Address: $multicastAddress:$multicastPort")
    }

    // ==================== Initialization Related Methods ====================

    /**
     * Update multicast port configuration
     */
    fun updateMulticastPort() {
        val settings = VSCodeJetBrainsSyncSettings.getInstance(project)
        val newPort = if (settings.state.useCustomPort) {
            settings.state.customPort
        } else {
            3000 // Default port for multicast when auto-detection is used
        }
        
        if (newPort != multicastPort) {
            log.info("Multicast port configuration changed: $multicastPort -> $newPort")

            // Update port
            multicastPort = newPort

            // If auto-reconnect is currently enabled, restart connection
            if (this.autoReconnect.get()) {
                this.restartConnection();
            }
        }
    }

    // ==================== Public Interface Methods ====================

    /**
     * Set connection state callback
     */
    fun setConnectionCallback(callback: ConnectionCallback) {
        this.connectionCallback = callback
    }

    /**
     * Toggle auto-reconnect state
     */
    fun toggleAutoReconnect() {
        connectionLock.lock()
        try {
            val currentState = autoReconnect.get()
            val newState = !currentState

            if (!autoReconnect.compareAndSet(currentState, newState)) {
                log.warn("Auto-reconnect state has been modified by another thread, operation cancelled")
                return
            }

            log.info("Multicast sync state toggled to: ${if (newState) "enabled" else "disabled"}")

            if (!newState) {
                disconnectAndCleanup()
                log.info("Multicast sync has been disabled")
            } else {
                connectMulticast()
                log.info("Multicast sync has been enabled, starting connection...")
            }
        } finally {
            connectionLock.unlock()
        }
    }

    /**
     * Restart connection
     */
    fun restartConnection() {
        log.info("Manually restarting multicast connection")
        disconnectAndCleanup()
        if (autoReconnect.get()) {
            connectMulticast()
        }
    }

    // ==================== Connection Management Methods ====================

    /**
     * Connect to multicast group
     */
    private fun connectMulticast() {
        if (isShutdown.get() || !autoReconnect.get() || connectionState.get() != ConnectionState.DISCONNECTED) {
            return
        }

        executorService.submit {
            try {
                if (!autoReconnect.get() || isShutdown.get()) {
                    return@submit
                }

                if (!connectionState.compareAndSet(ConnectionState.DISCONNECTED, ConnectionState.CONNECTING)) {
                    log.info("Connection state is not DISCONNECTED, skipping connection attempt")
                    return@submit
                }

                setConnectionState(ConnectionState.CONNECTING)
                log.info("Connecting to multicast group...")

                // Clean up existing connection
                cleanUp()

                // Find available network interface
                networkInterface = findAvailableNetworkInterface()
                if (networkInterface == null) {
                    throw RuntimeException("No available network interface found")
                }

                log.info("Using network interface: ${networkInterface!!.displayName}")

                // Create multicast socket
                multicastSocket = MulticastSocket(multicastPort)
                multicastSocket!!.reuseAddress = true
                multicastSocket!!.networkInterface = networkInterface

                // Join multicast group
                group = InetSocketAddress(InetAddress.getByName(multicastAddress), multicastPort)
                multicastSocket!!.joinGroup(group, networkInterface)

                setConnectionState(ConnectionState.CONNECTED)
                log.info("Successfully joined multicast group: $multicastAddress:$multicastPort")

                // Start message receiver thread
                startMessageReceiver()

            } catch (e: Exception) {
                log.warn("Failed to connect to multicast group: ${e.message}", e)
                handleConnectionError()
            }
        }
    }

    /**
     * Find available network interface
     */
    private fun findAvailableNetworkInterface(): NetworkInterface? {
        try {
            val interfaces = NetworkInterface.getNetworkInterfaces()
            if (!interfaces.hasMoreElements()) {
                log.error("No network interfaces found in the system")
                return null
            }

            val availableInterfaces = mutableListOf<String>()
            
            // Prioritize loopback interface to ensure local machine communication only
            for (netInterface in interfaces) {
                val interfaceInfo = "Interface: ${netInterface.displayName}, Loopback: ${netInterface.isLoopback}, Up: ${netInterface.isUp}, Multicast: ${netInterface.supportsMulticast()}"
                availableInterfaces.add(interfaceInfo)
                
                if (netInterface.isLoopback &&
                    netInterface.isUp &&
                    netInterface.supportsMulticast()
                ) {
                    log.info("Using loopback network interface: ${netInterface.displayName}")
                    return netInterface
                }
            }

            log.warn("Loopback interface not available, available interface list: ${availableInterfaces.joinToString("; ")}")
            
            // Reset enumerator
            val interfaces2 = NetworkInterface.getNetworkInterfaces()
            
            // If loopback interface is not available, try using any available interface as fallback
            for (netInterface in interfaces2) {
                if (netInterface.isUp && netInterface.supportsMulticast()) {
                    // More relaxed address check - try using as long as there's an IP address
                    val addresses = netInterface.inetAddresses
                    while (addresses.hasMoreElements()) {
                        val address = addresses.nextElement()
                        // Exclude IPv6 addresses (if multicast address is IPv4)
                        if (address is Inet4Address) {
                            log.info("Using network interface: ${netInterface.displayName}, Address: ${address.hostAddress}")
                            return netInterface
                        }
                    }
                }
            }

            // Final fallback: try using any enabled interface even if it doesn't support multicast
            log.warn("No multicast-supporting interface found, trying any enabled interface")
            val interfaces3 = NetworkInterface.getNetworkInterfaces()
            for (netInterface in interfaces3) {
                if (netInterface.isUp && netInterface.inetAddresses.hasMoreElements()) {
                    log.warn("Fallback using network interface: ${netInterface.displayName} (may not support multicast)")
                    return netInterface
                }
            }

            log.error("No available network interface found, checked system interfaces: ${availableInterfaces.joinToString("; ")}")
            return null

        } catch (e: Exception) {
            log.warn("Error occurred while finding network interface: ${e.message}", e)
            return null
        }
    }

    /**
     * Start message receiver thread
     */
    private fun startMessageReceiver() {
        receiverThread = thread(name = "Multicast-Message-Receiver") {
            val buffer = ByteArray(maxMessageSize)

            while (!isShutdown.get() && isConnected()) {
                try {
                    val packet = DatagramPacket(buffer, buffer.size)
                    multicastSocket?.receive(packet)

                    if (packet.length > 0) {
                        val message = String(packet.data, 0, packet.length, Charsets.UTF_8)
                        handleReceivedMessage(message)
                    }

                } catch (e: SocketTimeoutException) {
                    // Timeout is normal, continue loop
                    continue
                } catch (e: Exception) {
                    if (!isShutdown.get()) {
                        log.warn("Error occurred while receiving multicast message: ${e.message}", e)
                        handleConnectionError()
                        break
                    }
                }
            }

            log.info("Message receiver thread has exited")
        }
    }

    // ==================== Message Processing Methods ====================

    /**
     * Handle received message
     */
    private fun handleReceivedMessage(message: String) {
        try {
            messageProcessor.handleMessage(message)
        } catch (e: Exception) {
            log.warn("Error occurred while handling received message: ${e.message}", e)
        }
    }


    /**
     * Send message to multicast group
     */
    fun sendMessage(messageWrapper: MessageWrapper): Boolean {
        if (!isConnected() || !autoReconnect.get()) {
            log.warn("Currently not connected, discarding message: ${messageWrapper.toJsonString()}")
            return false
        }

        return try {
            val messageString = messageWrapper.toJsonString()
            val messageBytes = messageString.toByteArray(Charsets.UTF_8)

            if (messageBytes.size > maxMessageSize) {
                log.warn("Message too large, cannot send: ${messageBytes.size} bytes")
                return false
            }

            val packet = DatagramPacket(
                messageBytes,
                messageBytes.size,
                InetAddress.getByName(multicastAddress),
                multicastPort
            )

            multicastSocket?.send(packet)
            log.info("Sent multicast message content: $messageString")
            true

        } catch (e: Exception) {
            log.warn("Failed to send multicast message: ${e.message}", e)
            handleConnectionError()
            false
        }
    }

    // ==================== State Management Methods ====================

    /**
     * Handle connection error
     */
    private fun handleConnectionError() {
        setConnectionState(ConnectionState.DISCONNECTED)

        if (autoReconnect.get() && !isShutdown.get()) {
            executorService.submit {
                try {
                    Thread.sleep(5000) // Wait 5 seconds before reconnecting
                    if (autoReconnect.get() && !isShutdown.get()) {
                        connectMulticast()
                    }
                } catch (e: InterruptedException) {
                    // Thread was interrupted, exit
                }
            }
        }
    }

    /**
     * Set connection state and trigger callback
     */
    private fun setConnectionState(state: ConnectionState) {
        if (connectionState.get() == state) {
            return
        }

        connectionState.set(state)

        when (state) {
            ConnectionState.CONNECTED -> connectionCallback?.onConnected()
            ConnectionState.CONNECTING -> connectionCallback?.onReconnecting()
            ConnectionState.DISCONNECTED -> connectionCallback?.onDisconnected()
        }
    }

    // ==================== Resource Cleanup Methods ====================

    /**
     * Disconnect and clean up resources
     */
    fun disconnectAndCleanup() {
        cleanUp()
        setConnectionState(ConnectionState.DISCONNECTED)
    }

    /**
     * Clean up resources
     */
    private fun cleanUp() {
        connectionLock.lock()
        try {
            // Stop receiver thread
            receiverThread?.interrupt()
            receiverThread = null

            // Leave multicast group
            if (multicastSocket != null && group != null && networkInterface != null) {
                try {
                    multicastSocket!!.leaveGroup(group, networkInterface)
                    log.info("Left multicast group")
                } catch (e: Exception) {
                    log.warn("Error occurred while leaving multicast group: ${e.message}")
                }
            }

            // Close socket
            multicastSocket?.close()
            multicastSocket = null
            group = null
            networkInterface = null

        } finally {
            connectionLock.unlock()
        }
    }

    /**
     * Clean up resources
     */
    fun dispose() {
        log.info("Starting to clean up multicast manager resources")

        isShutdown.set(true)
        autoReconnect.set(false)

        // Clean up connection
        disconnectAndCleanup()

        // Close thread pool
        executorService.shutdown()
        try {
            if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
                executorService.shutdownNow()
            }
        } catch (e: InterruptedException) {
            executorService.shutdownNow()
        }

        log.info("Multicast manager resource cleanup completed")
    }

    // ==================== Status Query Methods ====================

    fun isConnected(): Boolean = connectionState.get() == ConnectionState.CONNECTED
    fun isAutoReconnect(): Boolean = autoReconnect.get()
    fun isConnecting(): Boolean = connectionState.get() == ConnectionState.CONNECTING
    fun isDisconnected(): Boolean = connectionState.get() == ConnectionState.DISCONNECTED
    fun getConnectionState(): ConnectionState = connectionState.get()
} 

File: vscode-extension/src/MulticastManager.ts
import * as dgram from 'dgram';
import * as vscode from 'vscode';
import {ConnectionCallback, ConnectionState, MessageWrapper} from './Type';
import {Logger} from './Logger';
import {MessageProcessor} from './MessageProcessor';
import {LocalIdentifierManager} from './LocalIdentifierManager';


/**
 * Multicast manager
 * Responsible for sending and receiving UDP multicast messages, implementing decentralized editor synchronization
 */
export class MulticastManager {
    // === Core Dependencies ===
    private readonly logger: Logger;
    private readonly messageProcessor: MessageProcessor;

    // === Network Configuration ===
    private readonly multicastAddress = '224.0.0.1'; // Local link multicast address, local machine communication only
    private multicastPort: number; // Multicast port (read from configuration)
    private readonly maxMessageSize = 8192; // Maximum message size (8KB)

    // === Connection State ===
    private socket: dgram.Socket | null = null;
    private connectionState: ConnectionState = ConnectionState.DISCONNECTED;
    private autoReconnect = false;
    private connectionCallback: ConnectionCallback | null = null;
    private isShutdown = false;

    // === Timers ===
    private reconnectTimer: NodeJS.Timeout | null = null;
    private cleanupTimer: NodeJS.Timeout | null = null;

    constructor(logger: Logger, messageProcessor: MessageProcessor) {
        this.logger = logger;
        this.messageProcessor = messageProcessor;
        this.multicastPort = vscode.workspace.getConfiguration('vscode-jetbrains-sync').get('port', 3000);

        this.logger.info(`Initializing multicast manager - Address: ${this.multicastAddress}:${this.multicastPort}`);

        this.setupConfigurationListener();
    }

    // ==================== Initialization Related Methods ====================

    /**
     * Set up configuration listener
     */
    private setupConfigurationListener(): void {
        vscode.workspace.onDidChangeConfiguration((event) => {
            if (event.affectsConfiguration('vscode-jetbrains-sync.port')) {
                this.updateMulticastPort();
            }
        });
    }

    /**
     * Handle configuration changes
     */
    private updateMulticastPort(): void {
        const newPort = vscode.workspace.getConfiguration('vscode-jetbrains-sync').get('port', 3000);
        if (newPort !== this.multicastPort) {
            this.logger.info(`Multicast port configuration changed: ${this.multicastPort} -> ${newPort}`);
            this.multicastPort = newPort;

            if (this.autoReconnect) {
                this.restartConnection();
            }
        }
    }

    // ==================== Public Interface Methods ====================

    /**
     * Set connection state callback
     */
    setConnectionCallback(callback: ConnectionCallback): void {
        this.connectionCallback = callback;
    }

    /**
     * Toggle auto-reconnect state
     */
    toggleAutoReconnect(): void {
        this.autoReconnect = !this.autoReconnect;
        this.logger.info(`Multicast sync state toggled to: ${this.autoReconnect ? 'enabled' : 'disabled'}`);

        if (!this.autoReconnect) {
            this.disconnectAndCleanup();
            this.logger.info('Multicast sync disabled');
        } else {
            this.connectMulticast();
            this.logger.info('Multicast sync enabled, starting connection...');
        }
    }

    /**
     * Restart connection
     */
    restartConnection(): void {
        this.logger.info('Manually restarting multicast connection');
        this.disconnectAndCleanup();
        if (this.autoReconnect) {
            this.connectMulticast();
        }
    }

    // ==================== Connection Management Methods ====================

    /**
     * Connect to multicast group
     */
    private connectMulticast(): void {
        if (this.isShutdown || !this.autoReconnect || this.connectionState !== ConnectionState.DISCONNECTED) {
            if (this.connectionState !== ConnectionState.DISCONNECTED) {
                this.logger.info('Connection state is not DISCONNECTED, skipping connection attempt');
            }
            return;
        }

        this.setConnectionState(ConnectionState.CONNECTING);
        this.logger.info('Connecting to multicast group...');

        try {
            this.cleanUp();
            this.createSocket();
            this.bindSocket();
        } catch (error) {
            this.logger.warn('Failed to create multicast connection:', error as Error);
            this.handleConnectionError();
        }
    }

    /**
     * Create UDP socket
     */
    private createSocket(): void {
        this.socket = dgram.createSocket({
            type: 'udp4',
            reuseAddr: true
        });

        this.setupSocketEventHandlers();
    }

    /**
     * Set up socket event handlers
     */
    private setupSocketEventHandlers(): void {
        if (!this.socket) return;

        this.socket.on('error', (error: Error) => {
            this.logger.warn('Multicast socket error:', error);
            this.handleConnectionError();
        });

        this.socket.on('message', (message: Buffer, rinfo: dgram.RemoteInfo) => {
            this.handleReceivedMessage(message.toString('utf8'));
        });

        this.socket.on('listening', () => {
            this.handleSocketListening();
        });
    }

    /**
     * Handle socket listening event
     */
    private handleSocketListening(): void {
        if (!this.socket) return;

        const address = this.socket.address();
        const isLoopback = address.address === '127.0.0.1' || address.address === '::1';
        const addressType = isLoopback ? 'loopback address' : 'non-loopback address';
        this.logger.info(`Multicast socket listening on ${addressType} ${address.address}:${address.port}`);

        this.joinMulticastGroup();
    }

    /**
     * Join multicast group
     */
    private joinMulticastGroup(): void {
        if (!this.socket) return;

        try {
            // Prioritize using loopback interface
            this.socket.addMembership(this.multicastAddress, '127.0.0.1');
            this.setConnectionState(ConnectionState.CONNECTED);
            this.logger.info(`Successfully joined multicast group (loopback interface): ${this.multicastAddress}:${this.multicastPort}`);
        } catch (error) {
            this.logger.warn('Failed to join multicast group (loopback interface), trying without specifying interface:', error as Error);
            this.tryJoinWithDefaultInterface();
        }
    }

    /**
     * Try to join multicast group using default interface
     */
    private tryJoinWithDefaultInterface(): void {
        if (!this.socket) return;

        try {
            this.socket.addMembership(this.multicastAddress);
            this.setConnectionState(ConnectionState.CONNECTED);
            this.logger.info(`Successfully joined multicast group (default interface): ${this.multicastAddress}:${this.multicastPort}`);
        } catch (error) {
            this.logger.warn('Failed to join multicast group completely:', error as Error);
            this.handleConnectionError();
        }
    }

    /**
     * Bind socket to port
     */
    private bindSocket(): void {
        if (!this.socket) return;

        try {
            // Prioritize binding to loopback address
            this.socket.bind(this.multicastPort, '127.0.0.1', () => {
                this.logger.info(`Bound to loopback address port: 127.0.0.1:${this.multicastPort}`);
            });
        } catch (bindError) {
            this.logger.warn('Failed to bind to loopback address, trying to bind to default address:', bindError as Error);
            this.tryBindToDefaultAddress();
        }
    }

    /**
     * Try to bind to default address
     */
    private tryBindToDefaultAddress(): void {
        if (!this.socket) return;

        try {
            this.socket.bind(this.multicastPort, () => {
                this.logger.info(`Bound to default address port: ${this.multicastPort}`);
            });
        } catch (error) {
            this.logger.warn('Failed to bind port completely:', error as Error);
            this.handleConnectionError();
        }
    }

    // ==================== Message Processing Methods ====================

    /**
     * Handle received message
     */
    private handleReceivedMessage(message: string): void {
        try {
            this.messageProcessor.handleMessage(message);
        } catch (error) {
            this.logger.warn('Error occurred while processing received message:', error as Error);
        }
    }


    /**
     * Send message to multicast group
     */
    sendMessage(messageWrapper: MessageWrapper): boolean {
        if (!this.isConnected() || !this.autoReconnect) {
            this.logger.warn(`Currently not connected, discarding message: ${messageWrapper.toJsonString()}`);
            return false;
        }

        try {
            const messageString = messageWrapper.toJsonString();
            const messageBuffer = Buffer.from(messageString, 'utf8');

            if (messageBuffer.length > this.maxMessageSize) {
                this.logger.warn(`Message too large, cannot send: ${messageBuffer.length} bytes`);
                return false;
            }

            this.socket!.send(
                messageBuffer,
                0,
                messageBuffer.length,
                this.multicastPort,
                this.multicastAddress,
                (error: Error | null) => {
                    if (error) {
                        this.logger.warn('Failed to send multicast message:', error);
                        this.handleConnectionError();
                    } else {
                        this.logger.info(`Sent multicast message content: ${messageString}`);
                    }
                }
            );

            return true;

        } catch (error) {
            this.logger.warn('Failed to send multicast message:', error as Error);
            this.handleConnectionError();
            return false;
        }
    }

    // ==================== State Management Methods ====================

    /**
     * Handle connection error
     */
    private handleConnectionError(): void {
        this.setConnectionState(ConnectionState.DISCONNECTED);

        if (this.autoReconnect && !this.isShutdown) {
            this.scheduleReconnect();
        }
    }

    /**
     * Schedule reconnection
     */
    private scheduleReconnect(): void {
        if (this.reconnectTimer) {
            clearTimeout(this.reconnectTimer);
        }

        this.logger.info('Will attempt to reconnect to multicast group in 5 seconds...');

        this.reconnectTimer = setTimeout(() => {
            if (this.autoReconnect && !this.isShutdown) {
                this.connectMulticast();
            }
        }, 5000);
    }

    /**
     * Set connection state and trigger callback
     */
    private setConnectionState(state: ConnectionState): void {
        if (this.connectionState === state) {
            return;
        }

        this.connectionState = state;

        switch (state) {
            case ConnectionState.CONNECTED:
                this.connectionCallback?.onConnected();
                break;
            case ConnectionState.CONNECTING:
                this.connectionCallback?.onReconnecting();
                break;
            case ConnectionState.DISCONNECTED:
                this.connectionCallback?.onDisconnected();
                break;
        }
    }

    // ==================== Resource Cleanup Methods ====================

    /**
     * Disconnect and clean up resources
     */
    disconnectAndCleanup(): void {
        this.cleanUp();
        this.setConnectionState(ConnectionState.DISCONNECTED);
    }

    /**
     * Clean up resources
     */
    private cleanUp(): void {
        // Clear timers
        if (this.reconnectTimer) {
            clearTimeout(this.reconnectTimer);
            this.reconnectTimer = null;
        }

        // Close socket
        if (this.socket) {
            try {
                this.socket.dropMembership(this.multicastAddress);
                this.logger.info('Left multicast group');
            } catch (error) {
                this.logger.warn('Error occurred while leaving multicast group:', error as Error);
            }

            this.socket.close(() => {
                this.logger.info('Multicast socket closed');
            });
            this.socket = null;
        }
    }

    /**
     * Clean up resources
     */
    dispose(): void {
        this.logger.info('Starting to clean up multicast manager resources');

        this.isShutdown = true;
        this.autoReconnect = false;

        // Clean up connection
        this.disconnectAndCleanup();

        // Clear cleanup timer
        if (this.cleanupTimer) {
            clearInterval(this.cleanupTimer);
            this.cleanupTimer = null;
        }

        this.logger.info('Multicast manager resource cleanup completed');
    }

    // ==================== Status Query Methods ====================

    isConnected(): boolean {
        return this.connectionState === ConnectionState.CONNECTED;
    }

    isAutoReconnect(): boolean {
        return this.autoReconnect;
    }

    isConnecting(): boolean {
        return this.connectionState === ConnectionState.CONNECTING;
    }

    isDisconnected(): boolean {
        return this.connectionState === ConnectionState.DISCONNECTED;
    }

    getConnectionState(): ConnectionState {
        return this.connectionState;
    }

}